src/functions.js

```diff
@@ -110,18 +110,35 @@ var _ = Mavo.Functions = {
 		return ret === null || !id? null : decodeURIComponent(ret[1]) || "";
 	},
 
-	// TODO return first/last non-null?
 	first: (n, arr) => {
 		if (arr === undefined) {
 			arr = n;
 			n = undefined;
 		}
 
 		if (!Array.isArray(arr)) {
-			return n && n > 1? [arr] : arr;
+			return n ? [arr] : arr;
+		}
+
+		if (n >= 0) {
+			var ret = [];
+			n = Math.floor(n);
+
+			for (var i = 0; i<arr.length && ret.length<n; i++) {
+				if (Mavo.value(arr[i]) !== null) {
+					ret.push(arr[i]);
+				}
+			}
+
+			return ret;
+		} 
+		else if (n < 0) {
+			return _.last(Math.abs(n), arr);
+		}
+		else { // n is undefined
+			return _.first(1, arr)[0];
```

Comment:
This will return `undefined` if the array has no non-null elements.
I didn't necessarily mean recursing btw, doing it in the main part of the code would be fine. 
You can get rid of negatives early, then define a variable that is `n === undefined? 1 : n` and run the loop based on that instead of n, then check if `n === undefined` at the end and return the first element if there is one or `null` otherwise.